File: gfx2Imp/java/Canvas2Peer.java
//
// Copyright (c) 2011, chunquedong
// Licensed under the Academic Free License version 3.0
//
// History:
//   2011-08-14  Jed Young  Creation
//

package fan.gfx2Imp;

import fan.sys.*;
import fan.fwt.*;
import fan.gfx.*;

import org.eclipse.swt.*;
import org.eclipse.swt.graphics.GC;
import org.eclipse.swt.graphics.Point;
import org.eclipse.swt.graphics.Rectangle;
import org.eclipse.swt.widgets.*;
import org.eclipse.swt.widgets.Widget;
import org.eclipse.swt.widgets.Canvas;
import org.eclipse.swt.layout.*;
import org.eclipse.swt.events.*;

public class Canvas2Peer extends CanvasPeer implements PaintListener
{

  public static Canvas2Peer make(Canvas2 self)
    throws Exception
  {
    Canvas2Peer peer = new Canvas2Peer();
    ((fan.fwt.Widget)self).peer = peer;
    peer.self = self;
    return peer;
  }

  public Widget create(Widget parent)
  {
    Canvas c = new Canvas((Composite)parent, SWT.NO_BACKGROUND);
    c.addPaintListener(this);
    return c;
  }

  public void paintControl(PaintEvent e)
  {
    FwtGraphics2 g = new FwtGraphics2(e);
    ((Canvas2)self).onPaint(g);
    ((Canvas2)self).onPaint2(g);
    g.dispose();
  }

  public void setCaret(Canvas2 self, long x, long y, long w, long h)
  {
    Caret caret = new Caret((Canvas)this.control(), SWT.NONE);
    caret.setBounds((int)x, (int)y, (int)w, (int)h);
  }
}

File: gfx2Imp/js/Canvas2Peer.js
//
// Copyright (c) 2009, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   12 Jun 09  Brian Frank  Creation
//

/**
 * CanvasPeer.
 */
fan.gfx2Imp.Canvas2Peer = fan.sys.Obj.$extend(fan.fwt.CanvasPeer);
fan.gfx2Imp.Canvas2Peer.prototype.$ctor = function(self) {}

fan.gfx2Imp.Canvas2Peer.prototype.setCaret = function(self, x, y, w, h)
{
  var c = this.getCanvas();
  if (!c) return;
  var cx = c.getContext("2d");
  try
  {
    cx.setCaretSelectionRect(c, x, y, w, h);
  }
  catch(e)
  {
    console.log("caret isn't supported");
  }
}

fan.gfx2Imp.Canvas2Peer.prototype.getCanvas = function()
{
  // short-circuit if not properly layed out
  var size = this.m_size
  if (size.m_w == 0 || size.m_h == 0) return;

  if (this.hasCanvas)
  {
    var div = this.elem;
    var c = div.firstChild;

    // remove old canvas if size is different
    if (c != null && (c.width != size.m_w || c.height != size.m_h))
    {
      div.removeChild(c);
      c = null;
    }

    // create new canvas if null
    if (c == null)
    {
      c = document.createElement("canvas");
      c.width  = size.m_w;
      c.height = size.m_h;
      div.appendChild(c);
    }

    return c;
  }
  else
  {
    console.log("don't suppert canvas2d");
    return null;
  }
}

fan.gfx2Imp.Canvas2Peer.prototype.sync = function(self)
{
  var c = this.getCanvas();
  if (!c) return;

  // repaint canvas using Canvas.onPaint callback
  var g = new fan.gfx2Imp.Graphics2();
  g.widget = self;
  g.paint(c, self.bounds(), function() { self.onPaint(g); self.onPaint2(g); })

  fan.fwt.WidgetPeer.prototype.sync.call(this, self);
}

